core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java

```diff
@@ -0,0 +1,171 @@
+package edu.wpi.grip.core.operations.composite;
+
+import com.google.common.eventbus.EventBus;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHints;
+import org.bytedeco.javacpp.BytePointer;
+import org.bytedeco.javacpp.IntPointer;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.bytedeco.javacpp.opencv_core.Mat;
+import static org.bytedeco.javacpp.opencv_imgcodecs.CV_IMWRITE_JPEG_QUALITY;
+import static org.bytedeco.javacpp.opencv_imgcodecs.imencode;
+
+/**
+ * Save JPEG files periodically to the local disk.
+ */
+public class SaveImageOperation implements Operation {
+
+    private final Logger logger = Logger.getLogger(SaveImageOperation.class.getName());
+
+    private final Object imageLock = new Object();
+    private final BytePointer imagePointer = new BytePointer();
+    private Optional<Thread> saveThread = Optional.empty();
+    private volatile boolean connected = false;
+    private long startTime = System.nanoTime();
+    private int numSteps = 0;
+    private SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd_HHmmss.SSS");
+    private String prefix = "./";
+    private String suffix = ".jpg";
+
+    /**
+     * Listens for incoming connections on port 1180 and writes JPEG data whenever there's a new frame.
+     */
+    private final Runnable runServer = () -> {
+        // Loop forever (or at least until the thread is interrupted).
+        logger.info("Starting image saver");
+        byte[] buffer = new byte[128 * 1024];
+        int bufferSize;
+        String prefix, suffix;
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                // Wait for the main thread to put a new image.
+                synchronized (imageLock) {
+                    imageLock.wait();
+
+                    // Copy the image data into a pre-allocated buffer, growing it if necessary
+                    bufferSize = imagePointer.limit();
+                    if (bufferSize > buffer.length) buffer = new byte[imagePointer.limit()];
+                    imagePointer.get(buffer, 0, bufferSize);
+                    prefix = this.prefix;
+                    suffix = this.suffix;
+                }
+            } catch (InterruptedException e) {
+                break;
+            }
+
+            // Write the file
+            try {
+                FileOutputStream out = new FileOutputStream(prefix + format.format(new Date()) + suffix);
```

Comment:
Try with resource would be more appropriate here.
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html